File: lw6/Korablyov_Anatoly/lw6_var2/MergeSort.cpp
#include "MergeSort.h"


CMergeSort::CMergeSort()
{
}


CMergeSort::~CMergeSort()
{
}
void StartCalcRank(CMergeSort * mat, std::vector<unsigned> & num)
{
	mat->StartMergeSort(num);
}

void ReplaceArrToTwoArr(const std::vector<unsigned> & nums, std::vector<unsigned> & firstArr, std::vector<unsigned> & secondArr)
{
	int nElems = nums.size() / 2;
	for (auto it : nums)
	{
		if (nElems < 1)
		{
			secondArr.push_back(it);
		}
		else
		{
			firstArr.push_back(it);
			nElems--;
		}
	}
}

std::vector<unsigned> CMergeSort::MergeTwoArray(const std::vector<unsigned> & firstArr, const std::vector<unsigned> & secondArr)
{
	std::vector<unsigned> resultArr;
	int firstIndex = 0;
	int secIndex = 0;
	while (resultArr.size() < firstArr.size() + secondArr.size())
	{
		if (firstIndex >= int(firstArr.size()))
		{
			for (size_t it = secIndex; it < secondArr.size(); it++)
			{
				resultArr.push_back(secondArr[it]);
			}
		}
		else if (secIndex >= int(secondArr.size()))
		{
			for (size_t it = firstIndex; it < firstArr.size(); it++)
			{
				resultArr.push_back(firstArr[it]);
			}
		}
		else
		{
			if (firstArr[firstIndex] <= secondArr[secIndex])
			{
				resultArr.push_back(firstArr[firstIndex]);
				firstIndex++;
			}
			else
			{
				resultArr.push_back(secondArr[secIndex]);
				secIndex++;
			}
		}
	}
	return resultArr;
}

void CMergeSort::StartMergeSort(std::vector<unsigned>& num)
{
	if (num.size() > 1)
	{
		std::vector<unsigned> firstArr;
		std::vector<unsigned> secondArr;
		ReplaceArrToTwoArr(num, firstArr, secondArr);

		StartMergeSort(firstArr);
		StartMergeSort(secondArr);

		std::vector<unsigned> resultArr = MergeTwoArray(firstArr, secondArr);

		num.clear();
		for (auto i : resultArr)
		{
			num.push_back(i);
		}
	}
}

std::vector<unsigned> CMergeSort::MergeSortParallel(const std::vector<unsigned>& num)
{
	if (num.size() > 1)
	{
		std::vector<unsigned> firstArr;
		std::vector<unsigned> secondArr;
		ReplaceArrToTwoArr(num, firstArr, secondArr);

		std::thread firstThread(StartCalcRank, this, firstArr);
		std::thread secThread(StartCalcRank, this, secondArr);
		firstThread.join();
		secThread.join();

		std::vector<unsigned> resultArr = MergeTwoArray(firstArr, secondArr);

		return resultArr;
	}
	else
	{
		return num;
	}
}

std::vector<unsigned> CMergeSort::MergeSortSequence(const std::vector<unsigned>& num)
{
	if (num.size() > 1)
	{
		std::vector<unsigned> firstArr;
		std::vector<unsigned> secondArr;
		ReplaceArrToTwoArr(num, firstArr, secondArr);

		StartMergeSort(firstArr);
		StartMergeSort(secondArr);

		std::vector<unsigned> resultArr = MergeTwoArray(firstArr, secondArr);

		return resultArr;
	}
	else
	{
		return num;
	}
}

File: lw7/Korablyov_Anatoly/lw7/MergeSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lw7
{
    class MergeSort
    {

        public static int[] MergeTwoArray(int[] firstArr, int[] secondArr)
        {
            int[] resultArr = new int[firstArr.Length + secondArr.Length];
            int firstIndex = 0;
            int secIndex = 0;
            int resIndex = 0;
            while (resultArr.Length < firstArr.Length + secondArr.Length)
            {
                if (firstIndex >= firstArr.Length)
                {
                    for (int it = secIndex; it < secondArr.Length; it++)
                    {
                        resultArr[resIndex] = secondArr[it];
                        resIndex++;
                    }
                }
                else if (secIndex >= secondArr.Length)
                {
                    for (int it = firstIndex; it < firstArr.Length; it++)
                    {
                        resultArr[resIndex] = firstArr[it];
                        resIndex++;
                    }
                }
                else
                {
                    if (firstArr[firstIndex] <= secondArr[secIndex])
                    {
                        resultArr[resIndex] = firstArr[firstIndex];
                        resIndex++;
                        firstIndex++;
                    }
                    else
                    {
                        resultArr[resIndex] = secondArr[secIndex];
                        resIndex++;
                        secIndex++;
                    }
                }
            }
            return resultArr;
        }

        public static void ReplaceArrToTwoArr(int[] num, int[] firstArr, int[] secondArr)
        {
            int nElems = num.Length / 2;

            int fIndex = 0;
            int sIndex = 0;
            for (int i = 0; i < num.Length; i++)
            {
                if (nElems < 1)
                {
                    secondArr[sIndex] = num[i];
                    sIndex++;
                }
                else
                {
                    firstArr[fIndex] = num[i];
                    fIndex++;
                    nElems--;
                }
            }
        }

        public static int[] MergeSortSequence(int[] num)
        {
            if (num.Length > 1)
            {
                int nElems = num.Length / 2;
                int[] firstArr = new int[nElems];
                int[] secondArr = new int[num.Length - nElems];

                ReplaceArrToTwoArr(num, firstArr, secondArr);

                int[] fNewArr = MergeSortSequence(firstArr);
                int[] sNewArr = MergeSortSequence(secondArr);

                int[] resArr = MergeTwoArray(fNewArr, sNewArr);
                return resArr;
            }
            return num;
        }

        public static int[] MergeSortParallel(int[] num)
        {
            if (num.Length > 1)
            {
                int nElems = num.Length / 2;
                int[] firstArr = new int[nElems];
                int[] secondArr = new int[num.Length - nElems];

                ReplaceArrToTwoArr(num, firstArr, secondArr);

                int[] fNewArr = new int[firstArr.Length];
                int[] sNewArr = new int[secondArr.Length];
                Task thread1 = Task.Factory.StartNew(() => fNewArr = MergeSortSequence(firstArr));
                Task thread2 = Task.Factory.StartNew(() => sNewArr = MergeSortSequence(secondArr));
                Task.WaitAll(thread1, thread2);

                int[] resArr = MergeTwoArray(fNewArr, sNewArr);
                return resArr;
            }
            return num;
        }

    }
}
